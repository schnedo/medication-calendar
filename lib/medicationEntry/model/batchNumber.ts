export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
