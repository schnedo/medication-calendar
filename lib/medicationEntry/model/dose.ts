export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
