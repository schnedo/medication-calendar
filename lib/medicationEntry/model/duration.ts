export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
