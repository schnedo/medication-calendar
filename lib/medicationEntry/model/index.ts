export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
