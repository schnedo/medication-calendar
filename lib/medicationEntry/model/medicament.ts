import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
