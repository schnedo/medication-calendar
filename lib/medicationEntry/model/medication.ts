import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
