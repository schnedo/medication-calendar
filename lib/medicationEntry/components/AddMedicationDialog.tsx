import {
  ChangeEventHandler,
  Dispatch,
  FocusEventHandler,
  PointerEventHandler,
  ReactElement,
  SetStateAction,
  useEffect,
  useState,
} from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormGroup,
  TextField,
} from "@material-ui/core";
import { ModalProps } from "@material-ui/core/Modal";
import { Medication } from "../model";
import { v4 } from "uuid";

function checkValidity<T>(
  value: T,
  validitySetter: Dispatch<SetStateAction<boolean>>,
): boolean {
  const valid = !!value;
  validitySetter(valid);
  return valid;
}

export interface AddMedicationDialogProps {
  open: boolean;
  onClose?: ModalProps["onClose"];
  onSubmit?: (medication: Medication) => Promise<void>;
  onAbort?: () => void;
}

export default function AddMedicationDialog({
  open,
  onClose,
  onSubmit,
  onAbort,
}: AddMedicationDialogProps): ReactElement {
  const [doseAmount, setDoseAmount] = useState("");
  const [doseValid, setDoseValid] = useState(true);
  const handleDoseChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setDoseAmount(value);
  const handleDoseBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setDoseValid);
  const handleDoseFocus: FocusEventHandler<HTMLInputElement> = () =>
    setDoseValid(true);

  const [name, setName] = useState("");
  const [nameValid, setNameValid] = useState(true);
  const handleNameChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setName(value);
  const handleNameBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setNameValid);
  const handleNameFocus: FocusEventHandler<HTMLInputElement> = () =>
    setNameValid(true);

  const [batchNumber, setBatchNumber] = useState("");
  const [batchNumberValid, setBatchNumberValid] = useState(true);
  const handleBatchNumberChange: ChangeEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => setBatchNumber(value);
  const handleBatchNumberBlur: FocusEventHandler<HTMLInputElement> = ({
    target: { value },
  }) => checkValidity(value, setBatchNumberValid);
  const handleBatchNumberFocus: FocusEventHandler<HTMLInputElement> = () =>
    setBatchNumberValid(true);

  useEffect(() => {
    if (!open) {
      const resetState = () => {
        setDoseAmount("");
        setDoseValid(true);
        setName("");
        setNameValid(true);
        setBatchNumber("");
        setBatchNumberValid(true);
      };
      resetState();
    }
  }, [open]);

  const handleClose: typeof onClose = (event, reason) => {
    onClose && onClose(event, reason);
  };
  const handleAbort = () => {
    onAbort && onAbort();
  };

  const handleSubmit:
    | PointerEventHandler<HTMLButtonElement>
    | undefined = onSubmit
    ? async () => {
        const doseValidity = checkValidity(doseAmount, setDoseValid);
        const nameValidity = checkValidity(name, setNameValid);
        const batchNumberValidity = checkValidity(
          batchNumber,
          setBatchNumberValid,
        );
        if (doseValidity && nameValidity && batchNumberValidity) {
          await onSubmit({
            id: v4(),
            medicament: {
              name,
              batchNumber: {
                number: parseInt(batchNumber),
              },
            },
            dose: {
              amount: parseInt(doseAmount),
            },
          });
        }
      }
    : undefined;

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle>Neue Medikation</DialogTitle>
      <DialogContent>
        <FormGroup>
          <TextField
            id={"doseInput"}
            label={"Dosis"}
            type={"number"}
            value={doseAmount}
            onChange={handleDoseChange}
            onBlur={handleDoseBlur}
            onFocus={handleDoseFocus}
            error={!doseValid}
            helperText={!doseValid && "Bitte Dosis eingeben"}
          />
          <TextField
            id={"nameInput"}
            label={"Name"}
            value={name}
            onChange={handleNameChange}
            onBlur={handleNameBlur}
            onFocus={handleNameFocus}
            error={!nameValid}
            helperText={!nameValid && "Bitte Name eingeben"}
          />
          <TextField
            id={"batchNumberInput"}
            label={"Chargenbezeichnung"}
            type={"number"}
            value={batchNumber}
            onChange={handleBatchNumberChange}
            onBlur={handleBatchNumberBlur}
            onFocus={handleBatchNumberFocus}
            error={!batchNumberValid}
            helperText={
              !batchNumberValid && "Bitte Chargenbezeichnung eingeben"
            }
          />
        </FormGroup>
      </DialogContent>
      <DialogActions>
        <Button color={"primary"} onClick={handleAbort}>
          Abbrechen
        </Button>
        <Button color={"primary"} onClick={handleSubmit}>
          Speichern
        </Button>
      </DialogActions>
    </Dialog>
  );
}
