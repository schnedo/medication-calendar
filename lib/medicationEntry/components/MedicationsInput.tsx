import { PointerEventHandler, ReactElement, useState } from "react";
import {
  Box,
  DialogContentText,
  IconButton,
  makeStyles,
  Menu,
  MenuItem,
} from "@material-ui/core";
import { Add } from "@material-ui/icons";
import MedicationDialog from "./MedicationDialog";
import { Medication } from "../model";
import { MedicationCard } from "./index";

const useStyles = makeStyles(() => ({
  medicationsAddContainer: {
    display: "flex",
    justifyContent: "space-between",
  },
}));

interface MedicationsListSelection {
  medication: Medication;
  pointerPosition: {
    clientX: number;
    clientY: number;
  };
}

export interface MedicationsInputProps {
  value: Medication[];
  onChange?: (medications: Medication[]) => void;
}

export default function MedicationsInput({
  value,
  onChange,
}: MedicationsInputProps): ReactElement {
  const [selection, setSelection] = useState<MedicationsListSelection | null>(
    null,
  );
  const closeMenu = () => setSelection(null);

  const [open, setOpen] = useState(false);
  const openDialog = () => setOpen(true);
  const closeDialog = () => {
    setOpen(false);
    closeMenu();
  };

  const handleAdd: PointerEventHandler<HTMLButtonElement> = () => {
    closeMenu();
    openDialog();
  };
  const handleDialogClose = () => closeDialog();
  const handleSubmit = async (medication: Medication) => {
    if (onChange) {
      const index = value.findIndex(({ id }) => id === medication.id);
      const newMedications =
        index === -1
          ? [...value, medication]
          : [...value.slice(0, index), medication, ...value.slice(index + 1)];
      onChange(newMedications);
    }
    closeDialog();
  };

  const handleEdit = () => openDialog();
  const handleDelete = () => {
    onChange &&
      selection &&
      onChange(
        value.filter((medication) => medication != selection.medication),
      );
    closeMenu();
  };

  const { medicationsAddContainer } = useStyles();
  return (
    <Box>
      <DialogContentText>Einträge</DialogContentText>
      {value.map((medication) => (
        <MedicationCard
          medication={medication}
          key={medication.id}
          onClick={({ clientX, clientY }) =>
            setSelection({ medication, pointerPosition: { clientX, clientY } })
          }
        />
      ))}
      <Box className={medicationsAddContainer}>
        <div />
        <IconButton onClick={handleAdd}>
          <Add />
        </IconButton>
      </Box>
      <Menu
        open={!!selection}
        onClose={closeMenu}
        anchorReference={"anchorPosition"}
        anchorPosition={
          selection
            ? {
                top: selection.pointerPosition.clientY,
                left: selection.pointerPosition.clientX,
              }
            : undefined
        }
      >
        <MenuItem onClick={handleEdit}>Bearbeiten</MenuItem>
        <MenuItem onClick={handleDelete}>Löschen</MenuItem>
      </Menu>
      <MedicationDialog
        open={open}
        medication={selection?.medication}
        onClose={handleDialogClose}
        onAbort={handleDialogClose}
        onSubmit={handleSubmit}
      />
    </Box>
  );
}
