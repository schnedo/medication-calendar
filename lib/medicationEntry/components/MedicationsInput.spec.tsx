/* eslint-disable sonarjs/no-identical-functions */
import { mock } from "@userlike/joke";
import { render } from "@testing-library/react";
import MedicationsInput from "./MedicationsInput";
import { Medication } from "../model";
import userEvent from "@testing-library/user-event";
import { AddMedicationDialogProps } from "./AddMedicationDialog";

const { default: MedicationCard } = mock(import("./MedicationCard"));
const { default: AddMedicationDialog } = mock(import("./AddMedicationDialog"));

const medication: Medication = {
  id: "2",
  medicament: {
    name: "name",
    batchNumber: {
      number: 1234,
    },
  },
  dose: {
    amount: 2,
  },
};

const value: Medication[] = [
  {
    id: "1",
    medicament: {
      name: "name",
      batchNumber: { number: 1 },
    },
    dose: {
      amount: 1,
    },
  },
];

describe("MedicationsInput", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    MedicationCard.mockReturnValue(<div>mocked MedicationCard</div>);
    AddMedicationDialog.mockReturnValue(<div>mocked AddMedicationDialog</div>);
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    const { container } = render(<MedicationsInput value={value} />);

    expect(container).toMatchSnapshot();
  });

  it("should call open AddMedicationDialog when clicked", async () => {
    expect.hasAssertions();

    const { getByRole } = render(<MedicationsInput value={value} />);

    const addButton = getByRole("button");

    userEvent.click(addButton);

    expect(AddMedicationDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: true }),
      {},
    );
  });

  it("should call onChange handler when AddMedicationDialog submits", async () => {
    expect.hasAssertions();
    const handleChange = jest.fn();
    const mockDialog = ({ onSubmit }: AddMedicationDialogProps) => (
      <div
        data-testid={"testid"}
        onClick={() => onSubmit && onSubmit(medication)}
      >
        mocked
      </div>
    );
    AddMedicationDialog.mockImplementation(mockDialog);

    const { getByRole, getByTestId } = render(
      <MedicationsInput value={value} onChange={handleChange} />,
    );

    userEvent.click(getByRole("button"));
    expect(handleChange).not.toHaveBeenCalled();
    userEvent.click(getByTestId("testid"));
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(handleChange).toHaveBeenLastCalledWith([...value, medication]);
  });

  it("should close AddMedicationDialog when it submits", async () => {
    expect.hasAssertions();

    const mockDialog = ({ onSubmit }: AddMedicationDialogProps) => (
      <div
        data-testid={"testid"}
        onClick={() => onSubmit && onSubmit(medication)}
      >
        mocked
      </div>
    );
    AddMedicationDialog.mockImplementation(mockDialog);

    const { getByRole, getByTestId } = render(
      <MedicationsInput value={value} />,
    );

    userEvent.click(getByRole("button"));
    userEvent.click(getByTestId("testid"));
    expect(AddMedicationDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });

  it("should close AddMedicationDialog when it aborts", async () => {
    expect.hasAssertions();

    const mockDialog = ({ onAbort }: AddMedicationDialogProps) => (
      <div data-testid={"testid"} onClick={() => onAbort && onAbort()}>
        mocked
      </div>
    );
    AddMedicationDialog.mockImplementation(mockDialog);

    const { getByRole, getByTestId } = render(
      <MedicationsInput value={value} />,
    );

    userEvent.click(getByRole("button"));
    userEvent.click(getByTestId("testid"));
    expect(AddMedicationDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });

  it("should set a close dialog when AddMedicationsInput onClose gets called", async () => {
    expect.hasAssertions();

    const mockDialog = ({ onClose }: AddMedicationDialogProps) => (
      <div
        data-testid={"testid"}
        onClick={() => onClose && onClose({}, "backdropClick")}
      >
        mocked
      </div>
    );
    AddMedicationDialog.mockImplementation(mockDialog);

    const { getByRole, getByTestId } = render(
      <MedicationsInput value={value} />,
    );

    userEvent.click(getByRole("button"));
    userEvent.click(getByTestId("testid"));
    expect(AddMedicationDialog).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );
  });
});
