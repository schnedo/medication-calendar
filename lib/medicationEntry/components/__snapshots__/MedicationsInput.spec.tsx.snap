// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicationsInput should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <p
      class="MuiTypography-root MuiDialogContentText-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Eintr√§ge
    </p>
    <div>
      mocked MedicationCard
    </div>
    <div
      class="MuiBox-root MuiBox-root-3 makeStyles-medicationsAddContainer-1"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorDisabled"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      mocked AddMedicationDialog
    </div>
  </div>
</div>
`;
